quinn-proto/src/connection/mod.rs

```diff
@@ -760,10 +784,104 @@ where
+                    if next_seq > self.retire_cid_seq && !ids_found {
+                        self.retire_cid_seq = next_seq;
+                    }
+                    // An endpoint MUST NOT
+                    // provide more connection IDs than the peer's limit.
+                    //
+                    // An endpoint MAY
+                    // send connection IDs that temporarily exceed a peer's limit if the
+                    // NEW_CONNECTION_ID frame also requires the retirement of any excess,
+                    // by including a sufficiently large value in the Retire Prior To field.
+                    // https://tools.ietf.org/html/draft-ietf-quic-transport-29#section-5.1.1
+                    // In this case, `retire prior to` must increase to accommodate new cid
+                    let must_increase_retire_cid =
+                        self.cids_active_seq.len() as u64 >= self.peer_params.issue_cids_limit();
+                    if must_increase_retire_cid
+                        && self.retire_cid_seq - current_seq < num_pushed_cids
+                    {
+                        self.retire_cid_seq = current_seq + num_pushed_cids;
+                    }
+
+                    // In case we already received RETIRE_CONNECTION_ID before
+                    let mut push_new_cid = false;
+                    for id in current_seq..self.retire_cid_seq {
+                        push_new_cid = push_new_cid || self.cids_active_seq.contains(&id);
+                    }
+
+                    if push_new_cid {
+                        if !self.state.is_closed() {
+                            trace!(
+                                "push a new cid to peer RETIRE_PRIOR_TO field {}",
+                                self.retire_cid_seq
+                            );
+                            self.endpoint_events
+                                .push_back(EndpointEventInner::NeedIdentifiers(
+                                    now,
+                                    num_pushed_cids,
+                                ));
+                        }
+                    } else {
+                        // We don't need to push new cid as peer has actively retired them already
+                        self.reset_cid_timer(None);
+                    }
+                }
             }
         }
     }
 
+    /// set timer for CID rotation
+    fn reset_cid_timer(&mut self, id: Option<CidTimeStamp>) {
+        // Remove any timestamp attached to a cid that is retired
+        while let Some(nc) = self.cids_timeline.front() {
+            if nc.sequence < self.retire_cid_seq {
+                self.cids_timeline.pop_front();
+            } else {
+                break;
+            }
+        }
+
+        if let Some(new_cid) = id {
+            if let Some(nc) = self.cids_timeline.front_mut() {
+                if new_cid.timestamp == nc.timestamp && new_cid.sequence > nc.sequence {
```

Comment:
Under what circumstances might `new_cid.sequence > nc.sequence` be false?

Reply:
This was possibly because i don't fully know the details of event handling in quinn. If all `ConnectionEvent` are managed by a VecDeque, we shouldn't need this condition.